### ISSUE TITLE
Single file Marko component

### ISSUE THREAD

**Issue by mlrawlings** [marko-js/marko#399]
We want to support defining a component in a single file.  To be clear, this would only be recommended for small components and demoing purposes.  You will still be able to have your template, styles, and component logic separated.

```html
<script>
module.exports = {
    handleButtonClick() {
    }
}
</script>

<style>
.foo {
}
</style>

<div class="foo">
</div>
```

---

### Other things we concidered:

<details><summary><b>Component-var</b></summary>

The "component" attribute with local `Component` var

```xml
<script component>
class Component {
    handleButtonClick() {}
}
</script>

<style component>
.foo {
}
</style>

<div class="foo">
</div>

<!-- Also allowed: -->

<div component=Component class=style('foo')>
</div>

<!-- Also allowed: -->

<div component="./component" class=style('foo')>
</div>
```

NOTES:

- The "component" attribute should be implicit on the non-`<style>`/non-`<script>` root HTML element if there is a `<script>` tag with a `component` attribute.
- `<script component>` replaces `<script marko-init>` ?
- What if there is `<script component>` with no `Component` var?
- `component.js` vs `Component.js` / `component` vs `Component`

</details>

<details><summary><b>component-attribute-export-default</b></summary>

The "component" attribute with `export default`

```xml
<script component>
export default {
    handleButtonClick() {}
}
</script>

<style component>
.foo {
}
</style>

<div class="foo">
</div>

<!-- Also allowed: -->
<script component>
module.exports = {
    handleButtonClick() {}
}
</script>
```

</details>

<details><summary><b>export-default-implicit</b></summary>

Optional "component" attribute when `export default`/`module.exports` detected

```xml
<script>
export default class {
    handleButtonClick() {
    }
}
</script>

<style component>
.foo {
}
</style>

<div class="foo">
</div>
```

</details>

<details><summary><b>vue style</b></summary>

Vue.js approach

```xml
<script>
export default class {
    handleButtonClick() {}
}
</script>

<style>
.foo {
}
</style>

<template>
    <div class="foo">
    </div>
</template>
```

NOTES:

- BAD: We are highjacking the `<template>` tag
- BAD: Template has to be wrapped in a `<template>` tag without much benefit
- BAD: Transition from a template-only component to a component with behavior requires wrapping the template with a `<template>` tag
- GOOD: Looks like a Web Component
- GOOD: Borrowing from Vue (or bad?)

</details>

<details><summary><b>autodetect-Component-var</b></summary>

Autodetect `<script>` with `Component` var/class

```xml
<script>
class Component {
    handleButtonClick() {}
}
</script>

<style component>
.foo {
}
</style>

<div class="foo">
</div>
```

NOTES:

Cons:
- What if a developer happens to want to render an inline `<script>` tag with a local variable named `Component`?
- When moving to an external file the 

</details>

<details><summary><b>component-tag wrapper</b></summary>

```xml
<component>
    <script>
    export default class {
        handleButtonClick() {}
    }
    </script>

    <style>
    .foo {
    }
    </style>
    
    <div class="foo">
    </div>
</component>
```

Pros:
- No random attributes or weird heuristics
- Using an export means the contents of `<component>` can be copied verbatim to an external file without any changes

Cons:
- Nice grouping, but what are you grouping it from?  There's likely nothing else in your file.
- Everything is forced to be nested a level deeper
- Moving to external files requires not only removing the script and style, but unwrapping the render logic.

</details>

<details><summary><b>component-tag as script</b></summary>

```xml
<component>
    export default class {
        handleButtonClick() {}
    }
</component>

<style>
    .foo {
    }
</style>
    
<div class="foo">
</div>
```

Pros:
- No random attributes or weird heuristics
- About as concise as we can get
- Transitioning to external files requires only removing the component and style
- Using an export means the contents of `<component>` can be copied verbatim to an external file without any changes
- Forces style to always be externalized? See question below.
    - If this is something we want, it makes it easy to move from a plain template, to a template with an embedded style, to a component with or without embedded logic, etc.

Cons: 
- Marko usually looks pretty good with plain `html` syntax highlighting, but you wouldn't get js syntax highlighting because we use a `<component>` instead of `<script>`. Marko specific syntax highlighting **would** be updated to give proper formatting, so maybe not a bug deal?
- Forces style to always be externalized? See question below.

Questions:
- Is it a good idea to always externalize an inline `<style>` tag?  I think yes.  At least when an asset taglib (like lasso) is installed.  And even if we didn't externalize, it would still work.  There just might be multiple `<style>` tags in the output if you render the template/component more than once.  CSS is global, so it doesn't make sense to allow input data to change global styles.  It works in the case of only one instance of the template being on the page, but otherwise it changes them all to the values of the last render.  You should really use the `style` attribute or a more advanced system if you want dynamic css.

</details>

---

### Goals:
- Not a lot of boilerplate
- Compatible with existing templates
- Easy transition from/to external files
- FUTURE: css-modules/scoped css

**Comment by mlrawlings** [comment]
@patrick-steele-idem @philidem This [gist](https://gist.github.com/mlrawlings/848a0b3c6d099bc1bffebc55e7f0acbd) makes a case for **component-tag as script** from the above list.  Open that section to read more about the pros/cons.

**Comment by patrick-steele-idem** [comment]
My vote is for `export-default-implicit` without the need for the `component` attribute on the `<style>` tag:

```xml
<script>
export default class {
    handleButtonClick() {
    }
}
</script>

<style>
.foo {
}
</style>

<div class="foo">
</div>
``` 

With that said, I think we need to have some more discussion around how to handle the `<style>` tag (e.g., is it always externalized or only in `index.marko` files?).

**Comment by mlrawlings** [comment]
The bulk of this is completed in https://github.com/marko-js/marko/commit/00b52ae5be30b9665bb289732adfd4bc985aa2e3.  Currently requires `w-bind` as it uses that to kick off looking for a top-level script tag.  Need to fix this along with #421.
